Adding support for plugins

vdwtanner opened the pull request:
Features:
* Scan for available plugins in %APPDATA%ColumnMode\Plugins
* Drop down menu for selecting a plugin to load
* PluginManager can call functions across all loaded plugins that have subscribed to the function. Currently:
  * OnSave, OnSaveAs
* Can call functions on a specific plugin using the Plugin class. Currently:
  * OnLoadCompleted
  * OnShutdown (not being called yet)
* Plugins can call callbacks in ColumnMode. Currently:
  * RegisterWindowClass, RegisterWindowClassEx, OpenWindow 

TODO:
- [ ] Unload plugins without closing ColumnMode
- [ ] Let Plugins define their own menu items
- [ ] Let plugins subscribe to hotkeys
- [ ] Config file to auto load plugins?
- [ ] Expand Plugin API

vdwtanner committed:
Stubbing out the beginnings of a plugin system

vdwtanner committed:
Mark Verify functions as inline

vdwtanner committed:
Add PluginManager global to Program.cpp

vdwtanner committed:
Upgrade to C++17

vdwtanner committed:
Begin Implementing Plugin loading

vdwtanner committed:
Plugins loading works. You can call functions on all plugins that have subscribed to a particular event.

Plugins still can't call back to ColumnMode.
Error messages are logged if a Plugin fails a call it said it supports.

vdwtanner committed:
Add a Sample Plugin project both for testing and demonstration

vdwtanner committed:
Hookup Save events

vdwtanner committed:
Scan for available plugins on startup

vdwtanner committed:
Basic menu system for Discovering Plugins and loading them

vdwtanner committed:
Add support for calling functions on specific plugins

vdwtanner committed:
Add ability to get function by name or ID

vdwtanner committed:
Begin working on WindowMaanger

vdwtanner committed:
Plugins can now create windows of their own within the ColumnMode process

vdwtanner committed:
OnSave* functions should pass plugin handle

clandrew commented:
Awesome, thanks for sending this!

clandrew commented on ColumnMode/ColumnMode.vcxproj:
```diff
@@ -23,33 +23,33 @@
     <ProjectGuid>{6546B3A0-08BA-43DD-A450-B2DE1262820A}</ProjectGuid>
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>ColumnMode</RootNamespace>
-    <WindowsTargetPlatformVersion>10.0.17134.0</WindowsTargetPlatformVersion>
```
Hate to do this, but can you remove this commit please? I'm not yet using VS2022.

vdwtanner replied:
Dropped the commit :)

clandrew commented on SamplePlugin/dllmain.cpp:
```diff
@@ -0,0 +1,74 @@
+    if (SUCCEEDED((*g_callbacks.pfnOpenWindow)(args, &g_hwnd)))
+    {
+        ShowWindow(g_hwnd, 1);
+    }
```
Minor thing- this doesn't return a value and falls of the end

vdwtanner replied:
Huh, surprised that compiled.

clandrew replied:
#justCPlusPlusThings

clandrew commented on ColumnMode/Program.cpp:
```diff
@@ -2188,6 +2209,62 @@ void OnPrint(WindowHandles windowHandles)
+	 constexpr int seperatorIndex = 1;
+	 for (int i = numItems; i > seperatorIndex ; i--)
+	 {
+		 HMENU m = (HMENU)GetMenuItemID(pluginMenu, i);
```
GetMenuItemID will fail (return -1) in the case where there are no plugins. 
Reason: there are 2 menu items
* Rescan
* Separator
i starts at 2, which isn't a valid item
It might be okay to DestroyMenu an HMENU handle of -1, but it's not good to rely on that. It'd be safer to either change the for loop or to check the return value of GetMenuItemID

vdwtanner replied:
Nice catch, thanks!

clandrew commented on ColumnMode/Verify.h:
```diff
@@ -1,12 +1,12 @@
-void VerifyHR(HRESULT hr)
+void inline VerifyHR(HRESULT hr)
 {
 	if (FAILED(hr))
```
I should probably get around to fixing this

clandrew replied:
Comment / no action requested

clandrew reviewed (changes requested):
It's very close to being merge able :) Thank you again! I tried out the sample plugin and it works great.

vdwtanner committed:
Add accelerator for SaveAs

vdwtanner committed:
Add plugin function call for OnOpen

vdwtanner commented:
Dropped VS2022 related changes and fixed the bugs you found. You may want to sanity check that it still builds properly in VS2017, but when I re-retargeted locally to 2022 it is building fine.

Also added Ctrl+Shift+S as a hotkey for SaveAs and another plugin function call for OnOpen.

clandrew commented:
The sample plugin project still depends on VS2022, because it uses the v143 toolset. If you're okay with it I can change it so not blocking

clandrew commented:
I grabbed the latest and hit this build error now

Build started...
1>------ Build started: Project: ColumnMode, Configuration: Debug x64 ------
1>PluginManager.cpp
1>d:\repos\columnmode\columnmode\pluginmanager.cpp(49): error C4996: 'std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>::copy': Call to 'std::basic_string::copy' with parameters that may be unsafe - this call relies on the caller to check that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. See documentation on how to use Visual C++ 'Checked Iterators'
1>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.14.26428\include\xstring(3348): note: see declaration of 'std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t>>::copy'
1>Done building project "ColumnMode.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========

Don't feel like blocking on this, will fix it when I get the chance

clandrew commented:
Just fixed up the last two things in main.
Thank you, @vdwtanner !! :)

clandrew commented:
I'm also a big fan of the Ctrl+Shift+S

vdwtanner commented:
Awesome, thanks Claire!
